tidy3d/components/autograd/utils.py

```diff
@@ -2,14 +2,37 @@
 
 import typing
 
-from autograd.tracer import getval
+import autograd.numpy as anp
+import numpy as np
+from autograd.tracer import getval, isbox
+
+
+def _is_traced_objary(x: typing.Any) -> bool:
+    """Check whether x is a traced object array."""
+    return isinstance(x, np.ndarray) and x.size > 0 and isbox(x.flat[0])
+
+
+def is_traced(x: typing.Any) -> bool:
+    """Check whether x is traced (contains an autograd Box)"""
+    return isbox(x) or _is_traced_objary(x)
 
 
 def get_static(x: typing.Any) -> typing.Any:
     """Get the 'static' (untraced) version of some value."""
     return getval(x)
 
 
+def get_box(x: typing.Any) -> typing.Any:
```

Comment:
`is_box` and `is_traced` look better to me here. Basically we're just adding a special case to handle `np.ndarray` with dtype=object and `ArrayBox` elements?

Reply:
Yeah exactly, this got refactored to the point where it probably doesn't make sense to have anymore :laughing:
If we don't do it here, we'll still need to handle object arrays somewhere else though.

Reply:
object arrays only pop up in the DataArray I think, so maybe we can just handle it there? unless it makes it easier to do it here.